keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,247 @@
+- CSI driver cannot share its limits with in-tree volume plugin e.g. when running pods with AWS EBS in-tree volumes and `ebs.csi.aws.com` CSI driver on the same node.
+- `node.status` size increases with each installed CSI driver. Node objects is big enough already.
+
+### Goals
+
+- User can run use PVs both with in-tree volume plugins and CSI and they will share their limits. There is only one scheduler predicate that handles both kind of volumes.
+
+- Existing predicates for in-tree volumes `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` are removed (with deprecation period).
+  - When both deprecated in-tree predicate and CSI predicate are enabled, only one of them does useful work and the other is NOOP to save CPU.
+
+- Increased CPU consumption as measured by [scheduler benchmark](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-scheduling/scheduler_benchmarking.md) is approved by sig-scheduling. There should be no performance regression in ideal case. 
+
+### Non-Goals
+
+
+## Proposal
+
+* Track volume limits for both in-tree volume plugins and CSI drivers in `CSINode` objects instead of `Node`.
+  * `Node` object is already big enough.
+  * Get rid of prefix + SHA for `ResourceName` of CSI volumes.
+  * In-tree volume plugin can share limits with CSI driver that uses the same storage backend.
+
+* Kubelet will create `CSINode` instance during initial node registration together with `Node` object.
+  * Limits of each in-tree volume plugin will be added to `CSINode.status.capacity` and `CSINode.status.allocatable`.
+    * Name of CSI driver corresponding to in-tree volume plugin will be used as `ResourceName` of these in-tree plugins, so the limits are shared between CSI driver and in-tree volume plugin in case both are running on the same node.
+    * If a CSI driver is registered for an in-tre volume plugin and it reports a different volume limit than in-tree volume plugin, the limit reported by CSI driver is used. (TODO: should kubelet `exit()`?)
+  * Kubelet will reconcile `CSINode.capacity`, overwriting any user's changes.
+  * User may change `CSINode.allocatable` to override volume plugin / CSI driver values, e.g. to "reserve" some attachment to the operating system.
+  * Kubelet will continue filling `Node.status.allocatable` and `Node.status.capacity` for both in-tree and CSI volumes during deprecation period. After the deprecation period, it will stop using them completely.
+    * User changes of `Node.status.allocatable` and `Node.status.capacity` will be ignored if `CSINode.status.allocatable` or `Node.status.capacity` is present.
```

Comment:
You mean ?
```suggestion
    * User changes of `Node.status.allocatable` and `Node.status.capacity` will be ignored if `CSINode.status.allocatable` or `CSINode.status.capacity` is present.
```

Reply:
Yeah, I reworked the section a bit. `Capacity` is gone and users can't change `capacity`  nor `allocatable` directly, they must use kubelet cmdline arguments.